jsonapi-generator/README.adoc

```diff
@@ -0,0 +1,287 @@
+is now capable of converting C++ objects from and to JSON.
+
+The current API is accessible via HTTP and unix socket, there is no
+authentication in both of them, so anyone having access to the HTTP server or to
+the unix socket can access the API without extra restrictions.
+Expecially for the HTTP API this is a big risk because also if the http server
+listen on 127.0.0.1 every application on the machine (even rogue javascript
+running on your web browser) can access that and for example on android it is
+not safe at all (because of that I implemented the unix socket access so at
+least in android API was reasonably safe) because of this.
+
+A second step to improve the API would be to implement some kind of API
+authentication mechanism (it would be nice that the mechanism is handled at API
+level and not at transport level so we can use it for any API trasport not just
+HTTP for example)
+
+The HTTP server used by libresapi is libmicrohttpd server that is very minimal,
+it doesn't provide HTTPS nor modern HTTP goodies, like server notifications,
+websockets etc. because the lack of support we have a token polling mechanism in
+libresapi to avoid polling for every thing but it is still ugly, so if we can
+completely get rid of polling in the API that would be really nice.
+I have done a crawl to look for a replacement and briefly looked at
+
+- https://www.gnu.org/software/libmicrohttpd/
+- http://wolkykim.github.io/libasyncd/
+- https://github.com/corvusoft/restbed
+- https://code.facebook.com/posts/1503205539947302/introducing-proxygen-facebook-s-c-http-framework/
+- https://github.com/cmouse/yahttp
+
+taking in account a few metrics like modern HTTP goodies support, license,
+platform support, external dependencies and documentation it seemed to me that
+restbed is the more appropriate.
+
+Another source of boilerplate code into libresapi is the mapping between JSON
+API requests and C++ API methods as an example you can look at this
+
+https://github.com/RetroShare/RetroShare/blob/v0.6.4/libresapi/src/api/ChatHandler.cpp#L158
+
+and this
+
+https://github.com/RetroShare/RetroShare/blob/v0.6.4/libresapi/src/api/ApiServer.cpp#L253
+
+The abstract logic of this thing is, when libreasapi get a request like
++/chat/initiate_distant_chat+ then call
++ChatHandler::handleInitiateDistantChatConnexion+ which in turn is just a
+wrapper of +RsMsgs::initiateDistantChatConnexion+ all this process is basically
+implemented as boilerplate code and would be unnecessary in a smarter design of
+the API because almost all the information needed is already present in the
+C++ API +libretroshare/src/retroshare+.
+
+So a third step to improve the JSON API would be to remove this source of
+boilerplate code by automatizing the mapping between C++ and JSON API call.
+
+This may result a little tricky as language parsing or other adevanced things
+may be required.
+
+Hope this dive is useful for you +
```

Comment:
you may want to remove these 3 lines ;-)

Reply:
Well that is left there just as historical record on how this idea came out ;)